src/mqtt.c

```diff
@@ -207,7 +208,14 @@ static void on_message (
     }
     vl.values_len = ds->ds_num;
 
-    payload = strdup ((void *) msg->payload);
+    payload = calloc (msg->payloadlen+1, sizeof (*payload));
+    if (payload == NULL)
+    {
+        ERROR ("mqtt plugin: calloc for payload buffer failed.");
+        return;
+    }
+    strncpy (payload, msg->payload, msg->payloadlen);
```

Comment:
Please use `memmove()` instead of `strncpy()` if the payload is not necessarily null terminated. Also, enforce null termination explicitly (relying on `calloc()` is too subtle to be easily readable):

```c
memmove (payload, msg->payload, msg->payloadlen);
payload[msg->payloadlen] = 0;
```

This should also fix the implicit `unsigned char *` to `char *` casting that @mfournier pointed out.